File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public boolean isPalindrome(String s, int low, int high) {
        while (low < high) {
            if (s.charAt(low++) != s.charAt(high--)) {
                return false;
            }
        }
        return true;
    }
    
    public void generate(int idx, String s, ArrayList<String> vec, List<List<String>> res) {
        if (idx == s.length()) {
            res.add(new ArrayList<>(vec));
            return;
        }
        for (int i = idx; i < s.length(); i++) {
            if (isPalindrome(s, idx, i)) {
                vec.add(s.substring(idx, i + 1));
                generate(i + 1, s, vec, res);
                vec.remove(vec.size() - 1);
            }
        }
    }
    
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        generate(0, s, new ArrayList<>(), res);
        return res;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool is_palindrome(string& s, int low, int high) {
        while (low < high) {
            if (s[low] != s[high]) {
                return false;
            }
            low++;
            high--;
        }
        return true;
    }
    
    void f(int i, string& s, vector<vector<string>>& res, int& n, vector<string>& ds) {
        if (i == n) {
            res.push_back(ds);
            return;
        }
        
        for (int j = i; j < n; j++) {
            if (is_palindrome(s, i, j)) {
                ds.push_back(s.substr(i, j - i + 1));
                f(j + 1, s, res, n, ds);
                ds.pop_back();
            }
        }
    }
    
    vector<vector<string>> partition(string s) {
        int n = s.size();
        
        vector<vector<string>> res;
        
        vector<string> ds;
        
        f(0, s, res, n, ds);
        
        return res;
    }
};